Handle more cases of undefined behaviour due to signed integer overflow or division by zero

jamienicol opened the pull request:
This is quite large so is in separate commits both for reviewability and future readability - please land with a rebase rather than squash!!

**Connections**
Fixes #6961 

**Description**
Signed arithmetic overflow is undefined behaviour as specified in the metal spec, and we have been informed that we cannot rely on it being safe in HLSL either. The simple cases (addition, subtraction, multiplication) are safe in SPIRV, and were handled in #6666 for metal, but still affect the HLSL backend. Additionally, some edge cases remain undefined behaviour for various backends:

* INT_MIN / -1
* INT_MIN % -1
* signed or unsigned / 0
* signed or unsigned % 0
* abs(INT_MIN)
* -INT_MIN

This patch series ensures all these cases are handled if necessary for the MSL, HLSL, and SPIRV backends, in line with the WGSL spec's requirements.

**Testing**
Inspected test snapshots and ensure they still pass validation. Manual testing that compute shader outputs correct values for various operations.

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

jamienicol committed:
[naga hlsl-out] Wrap int literals in constructor-style cast

There is no literal suffix in HLSL for the integer type. We can,
however, wrap integer literals in a constructor style cast. This
avoids ambiguity when passing literals to overloaded functions, which
we'll make use of in the subsequent patch in this PR.

jamienicol committed:
[naga hlsl-out] Avoid undefined behaviour for integer division, modulo, negation, and abs

Emit helper functions for MathFunction::Abs and UnaryOperator::Negate
with a signed integer scalar or vector operand. And for
BinaryOperator::Divide and BinaryOperator::Modulo with signed or
unsigned integer scalar or vector operands.

Abs and Negate are written to avoid signed integer overflow when the
operand equals INT_MIN. This is achieved by bitcasting the value to
unsigned, using the negation operator, then bitcasting the result back
to signed. As HLSL's bitcast functions asint() and asuint() only work
for 32-bit types, we only use this workaround in such cases.

Division and Modulo avoid undefined behaviour for INT_MIN / -1 and
divide-by-zero by using 1 for the divisor instead. Additionally we
avoid undefined behaviour when using the modulo operator on operands
of mixed signedness by using the equation from the WGSL spec, using
division, subtraction and multiplication, rather than HLSL's modulus
operator.

jamienicol committed:
[naga hlsl-out] Avoid undefined behaviour for signed integer addition, subtraction, and multiplication

Though not explicitly specified one way or the other, we have been
informed by DirectX engineers that signed integer overflow may be
undefined behaviour in some cases. To avoid this, we therefore bitcast
signed operands to unsigned prior to performing addition, subtraction,
or multiplication, then bitcast the result back to signed. As signed
types are represented as two's complement, this gives the correct
result whilst avoid any potential undefined behaviour.

Unfortunately HLSL's bitcast functions asint() and asuint() only work
for the 32-bit int and uint types. We therefore only apply this
workaround for 32-bit signed arithmetic. Support for other bit widths
could be added in the future, but extra care must be taken when
converting from unsigned to signed to avoid undefined or
implemented-defined behaviour.

jamienicol committed:
[naga msl-out] Avoid undefined behaviour for integer division, modulo, negation, and abs

Adds infrastructure to the MSL backend for emitting helper functions,
based upon the existing HLSL backend equivalent. Emit helper functions
for MathFunction::Abs and UnaryOperator::Negate with a signed integer
scalar or vector operand. And for BinaryOperator::Divide and
BinaryOperator::Modulo with signed or unsigned integer scalar or
vector operands.

Abs and Negate are written to avoid signed integer overflow when the
operand equals INT_MIN. This is achieved by bitcasting the value to
unsigned and using the negation operator, then bitcasting the result
back to signed.

Division and Modulo avoid undefined bevaviour for INT_MIN / -1 and
divide-by-zero by using 1 for the divisor instead. Additionally we
avoid undefined behaviour when using the modulo operator when either
or both operands are negative by using the equation from the WGSL
spec, using division, subtraction and multiplication, rather than
MSL's modulus operator.

Lastly, as the usage of the wrapper function for unary integer
negation makes the negation_avoids_prefix_decrement() testcase less
interesting, we extend it to additionally test negating floats.

jamienicol committed:
[naga spv-out] Avoid undefined behaviour for integer division and modulo

Integer division or modulo is undefined behaviour in SPIR-V when the
divisor is zero, or when the dividend is the most negative number
representable by the result type and the divisor is negative one.

This patch makes us avoid this undefined behaviour and instead ensures
we adhere to the WGSL spec in these cases: for divisions the
expression evaluates to the value of the dividend, and for modulos the
expression evaluates to zero.

Similarily to how we handle these cases for the MSL and HLSL backends,
prior to emitting each function we emit code for any helper functions
required by that function's expressions. In this case that is helper
functions for integer division and modulo. Then, when emitting the
actual function's body, if we encounter an expression which needs
wrapped we instead emit a function call to the helper.

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                        let level = crate::back::Level(1);
+                        writeln!(
+                            self.out,
+                            "{level}return val >= 0 ? val : asint(~asuint(val) + 1u);"
```
Doesn't HLSL support unary `-` on `uint`? C++ does. So this could just be 
```suggestion
                            "{level}return val >= 0 ? val : asint(-asuint(val));"
```

jamienicol replied:
ah unsigned negation never even occurred to me. Can't see anything in the respective specs, but the shaders appear to pass validation so presumably it's supported.

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                        writeln!(self.out, " val) {{")?;
+
+                        let level = crate::back::Level(1);
+                        writeln!(self.out, "{level}return asint(~asuint(val) + 1u);",)?;
```
Similarly:

```suggestion
                        writeln!(self.out, "{level}return asint(-asuint(val));",)?;
```

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+
+                match (
+                    op,
+                    func_ctx
+                        .resolve_type(expr_handle, &module.types)
+                        .scalar_kind(),
```
Isn't this just `expr_ty`?

jamienicol replied:
yes

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                        crate::BinaryOperator::Divide,
+                        Some(kind @ crate::ScalarKind::Sint | kind @ crate::ScalarKind::Uint),
+                    ) => {
+                        let left_wrapped_ty = match *func_ctx.resolve_type(left, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
+                        let right_wrapped_ty = match *func_ctx.resolve_type(right, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
```
Aren't these `resolve_type` calls exactly `left_ty` and `right_ty`, already computed above?

jamienicol replied:
yes again

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+        for (_, expression) in func_ctx.expressions.iter() {
+            if let crate::Expression::Unary { op, expr } = *expression {
+                let expr_ty = func_ctx.resolve_type(expr, &module.types);
+                let (scalar, vector_size) = match *expr_ty {
+                    crate::TypeInner::Scalar(scalar) => (scalar, None),
+                    crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                    _ => continue,
+                };
```
You repeat this match a lot. Pity the poor reader who has to check that these are all the same, and wonder why if they're not. How about turning it into an `impl TypeInner` utility function (`vector_size_and_scalar`, say) in `naga/src/proc/mod.rs`?

jamienicol replied:
done. the helpers in `proc/mod.rs` seem to have been recently moved to `proc/type_methods.rs` so that's where I put it

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                        let level = crate::back::Level(1);
+                        match kind {
+                            crate::ScalarKind::Sint => {
+                                let min = 2u64.pow(expr_ty.scalar_width().unwrap() as u32 * 8 - 1);
```
The `unwrap` is kind of a code smell: if you know it's not going to panic, then you must have done some prior match that you ought to have been able to get the data out of.

This is actually the negative of the minimum, so the name is a little confusing.

Suggestion: drop the explicit `-` in the format string and then use:

```suggestion
                                let min = -1_i64 << (expr_ty.scalar_width().unwrap() as u32 * 8 - 1);
```

That should get you the right minimum even for 64-bit integer types.

jimblandy commented on naga/src/back/hlsl/writer.rs:
```diff
@@ -2579,7 +2583,7 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                 crate::Literal::F32(value) => write!(self.out, "{value:?}")?,
                 crate::Literal::U32(value) => write!(self.out, "{value}u")?,
-                crate::Literal::I32(value) => write!(self.out, "{value}")?,
+                crate::Literal::I32(value) => write!(self.out, "int({value})")?,
```
Why is this necessary? This should have a comment, at least.

jamienicol replied:
done

jimblandy commented on naga/src/back/hlsl/writer.rs:
```diff
@@ -2719,16 +2749,40 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
+                write!(self.out, ")")?;
+            }
+
             // TODO: handle undefined behavior of BinaryOperator::Modulo
```
Can we delete this comment too?

jamienicol replied:
I believe we still need to handle division by zero for floating points

jimblandy replied:
Where does HLSL say floating-point division by zero is UB? It seems like they usually follow IEEE.

jimblandy replied:
This table is describing DXIL, and seems to imply that it's defined.

https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#fdiv

jimblandy replied:
More links: https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#3.1%20Floating%20Point%20Rules

jamienicol replied:
Sorry I meant to say modulo not division. But you might be right we can remove the comment, or at least have to update it. The modulus operator says it's undefined unless both operands are positive or negative. But we don't use the modulus operator, we use `fmod()`. This doesn't say anything about being undefined. But we do need to check whether it has the same semantics as WGSL..

jamienicol replied:
Removed the reference to undefined behaviour, but left the "TODO: if right == 0 return ?" part for now.

jimblandy reviewed (commented):
Still reviewing, some comments so far:

jimblandy commented on naga/src/back/hlsl/writer.rs:
```diff
@@ -26,6 +26,10 @@ pub(crate) const EXTRACT_BITS_FUNCTION: &str = "naga_extractBits";
 pub(crate) const INSERT_BITS_FUNCTION: &str = "naga_insertBits";
 pub(crate) const SAMPLER_HEAP_VAR: &str = "nagaSamplerHeap";
 pub(crate) const COMPARISON_SAMPLER_HEAP_VAR: &str = "nagaComparisonSamplerHeap";
+pub(crate) const ABS_FUNCTION: &str = "naga_abs";
+pub(crate) const DIV_FUNCTION: &str = "naga_div";
+pub(crate) const MOD_FUNCTION: &str = "naga_mod";
+pub(crate) const NEG_FUNCTION: &str = "naga_neg";
```
These need to be added to `back::hlsl::keywords::RESERVED`, so that they won't conflict with the user's identifiers.

jimblandy commented on naga/src/back/msl/writer.rs:
```diff
@@ -36,6 +35,10 @@ const RAY_QUERY_FUN_MAP_INTERSECTION: &str = "_map_intersection_type";
 pub(crate) const ATOMIC_COMP_EXCH_FUNCTION: &str = "naga_atomic_compare_exchange_weak_explicit";
 pub(crate) const MODF_FUNCTION: &str = "naga_modf";
 pub(crate) const FREXP_FUNCTION: &str = "naga_frexp";
+pub(crate) const ABS_FUNCTION: &str = "naga_abs";
+pub(crate) const DIV_FUNCTION: &str = "naga_div";
+pub(crate) const MOD_FUNCTION: &str = "naga_mod";
+pub(crate) const NEG_FUNCTION: &str = "naga_neg";
```
As with HLSL, you need to tell the namer about these.

jimblandy commented:
This also fixes #4385, I think.

jimblandy reviewed (commented):
Still reviewing, just posting some comments

jimblandy commented:
I like the infrastructure added to the Metal backend for this. Filed #7105 to take advantage of it more.

jimblandy commented:
I've fully reviewed the following commits and they look good to me:
- [naga msl-out] Avoid undefined behaviour for integer division, modulo, negation, and abs
- [naga hlsl-out] Wrap int literals in constructor-style cast
- [naga hlsl-out] Avoid undefined behaviour for signed integer addition, subtraction, and multiplication

jimblandy commented:
Filed #7109 for non-32-bit integer overflow UB in HLSl.

jimblandy committed:
[naga hlsl-out] Use `Scalar` associated constants in patterns.

jimblandy committed:
[hlsl-out] Clean up repetition in `write_mapped_math_functions`.

Since every `match` arm ends up looking up the type of the operation's
first argument, just do that once. This avoids a repetitive lookup for
`Abs`.

jimblandy committed:
[hlsl-out] Clarify comment on HLSL floating-point divide by zero.

Replace the link to the resolved WGSL spec issue about floating-point
division by zero (gpuweb/gpuweb#2798) with links to the Direct3D 11
functional specification (which Direct3D 12 inherits) and the DXIL
specification, explaining that HLSL does what WGSL wants here.

jimblandy commented:
I've fully reviewed the following commits and they look good to me:

> [naga hlsl-out] Avoid undefined behaviour for integer division, modulo, negation, and abs

jimblandy commented:
@jamienicol The commit messages are great and much appreciated! But for future PRs, there's a lot there that could serve (even) better as comments in the code, for example:

> Integer division or modulo is undefined behaviour in SPIR-V when the
divisor is zero, or when the dividend is the most negative number
representable by the result type and the divisor is negative one.

This wouldn't be amiss as a comment in the SPIR-V backend.

> This patch makes us avoid this undefined behaviour and instead ensures
we adhere to the WGSL spec in these cases: for divisions the
expression evaluates to the value of the dividend, and for modulos the
expression evaluates to zero.

This would be more useful as comments clarifying the behavior of the Naga IR operators: you'd describe it as Naga IR's semantics, and then note that this aligns with WGSL's operators.

Generally I try to put everything possible into comments in the code, and then let the commit message cite those, and focus on the changes to behavior.

jamienicol committed:
[naga msl-out] Document the need for wrapper functions for integer division, modulo, abs(), and unary negation

Explain we need the wrapper functions not just to avoid undefined
behaviour (or unspecified in the case of division), but also to ensure
we adhere to the WGSL spec.

jamienicol committed:
[naga hlsl-out] Document the need for wrapper functions for integer division, modulo, abs(), and unary negation

Explain we need the wrapper functions not just to avoid undefined
behaviour, but also to ensure we adhere to the WGSL spec. Additionally
link to issue #7109 in cases where our workaround needs follow-up work
for non-32-bit integer types.

jamienicol commented:
> @jamienicol The commit messages are great and much appreciated! But for future PRs, there's a lot there that could serve (even) better as comments in the code, for example:
> 
> > Integer division or modulo is undefined behaviour in SPIR-V when the
> > divisor is zero, or when the dividend is the most negative number
> > representable by the result type and the divisor is negative one.
> 
> This wouldn't be amiss as a comment in the SPIR-V backend.

This is actually commented [here](https://github.com/gfx-rs/wgpu/pull/7012/commits/64b0e1184ed588f2fb4d8ec39f5dd08bc743a471#diff-ec3f6fa5321c2bda7bfc667729f56f28c33e067454328ad22372fe2218709549R328)

But I think the comment is absolutely fair and I've pushed extra patches to the PR to add comments for the HLSL and MSL changes.

> 
> > This patch makes us avoid this undefined behaviour and instead ensures
> > we adhere to the WGSL spec in these cases: for divisions the
> > expression evaluates to the value of the dividend, and for modulos the
> > expression evaluates to zero.
> 
> This would be more useful as comments clarifying the behavior of the Naga IR operators: you'd describe it as Naga IR's semantics, and then note that this aligns with WGSL's operators.

Do you mean we should add comments to eg `BinaryOperator::Divide` etc declaration, in naga/src/lib.rs?

jimblandy committed:
[naga spv-out] Factor out wrapped divide/module generation.

Move the code to generate the definition of an overflow-safe
divide/modulo SPIR-V function into its own Rust function, to reduce
indentation and clarify influences. This commit isn't intended to
cause any change in behavior.

jimblandy commented:
> Do you mean we should add comments to eg `BinaryOperator::Divide` etc declaration, in naga/src/lib.rs?

Yeah, although I admit none of the other operators are documented, so it's going to look a little weird. May be a bigger project than belongs in this PR.

jimblandy commented on naga/src/back/spv/writer.rs:
```diff
@@ -305,6 +311,243 @@ impl Writer {
+                    lhs_eq_int_min_id,
+                    rhs_eq_neg_one_id,
+                ));
+                let rhs_eq_zero_or_lhs_eq_int_min_and_rhs_eq_neg_one_id = self.id_gen.next();
```
`dabbrev-expand` to the rescue!

jimblandy commented:
I've fully reviewed this commit and it looks fine to me:

> [naga spv-out] Avoid undefined behaviour for integer division and modulo

jimblandy reviewed (approved):
Okay, this all looks fantastic! Thank you!
